lib/src/main/kotlin/com/swmansion/starknet/data/serializers/TypedDataTypeBaseSerializer.kt

```diff
@@ -0,0 +1,18 @@
+package com.swmansion.starknet.data.serializers
+
+import com.swmansion.starknet.data.TypedData
+import com.swmansion.starknet.data.TypedData.TypeBase
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.json.*
+
+internal object TypedDataTypeBaseSerializer : JsonContentPolymorphicSerializer<TypeBase>(TypeBase::class) {
+    override fun selectDeserializer(element: JsonElement): DeserializationStrategy<out TypeBase> {
+        val type = element.jsonObject["type"]?.jsonPrimitive?.content
+
+        return when (type) {
+            "merkletree" -> TypedData.MerkleTreeType.serializer()
+            is String -> TypedData.Type.serializer()
+            else -> throw IllegalArgumentException("Input element does not contain mandatory field 'type'")
```

Comment:
It may contain this field, but we are just not checking for it. " .. contains unknown type field: <type>" may be a better message

Reply:
> but we are just not checking for it.

Do you mean that that the field is present but has a different name? This message meant to show that the field named `type` is missing.

Reply:
```
val type = element.jsonObject["type"]?.jsonPrimitive?.content
```
`type` here is always `String?`, so the only way the else condition is reached is by `type` being null. I can add an extra check if you think it should be there though.

Reply:
perhaps, but just modifying the message will be enough imo, to include an information that it either does not exist or is unknown. you never know when allowed types change, my thinking here is maybe it is better to be more explicit with the error message